### PR TITLE
Update CI to allow newer Python versions

### PR DESCRIPTION
**Pull request by eddiebergman**
(WIP)
Example of the output for the (github workflows) here:
* [Tests](https://github.com/eddiebergman/GraKeL/actions/runs/3041633008)
* [Wheels](https://github.com/eddiebergman/GraKeL/actions/runs/3041633006)

### TODOS
* I'm used to using `pytest` and `coverage` and having [`codecov`](https://about.codecov.io/tool/github-actions/) be the CI part that reports back on the PR. I'm not sure what to do here in this case. `codecov` is free for open-source repos though. Alternatively, a plain text version of the report could be printed and also left as a bot comment on the PR (if it's a PR).
* GraKeL seems to encounter its first backwards breaking change with Python 3.10, namely imports from `collections` were deprecated a while back and are now available in the `abc.collections` or `typing` module of python. See this [deprecation warning below the table](https://docs.python.org/3.9/library/collections.html) in the Python3.9 docs.
* Some slight fixes to do with the different builds, just a step by step debug and fix process
* The `where=grakel/tests` in `setup.cfg` seems to confuse `nosetest` when run on windows, trying some workaround or even just let auto-discovery take charge :'(

### Changes
* Drop Travis and AppVeyor, instead use Github Actions and [cibuildwheel](https://github.com/pypa/cibuildwheel) to support linux/mac/windows.
* Unfortunatly, it seems cibuildwheel is relatively indate with [pythons end of life schedule](https://endoflife.date/python). This means that the "easy" setup means new releases will not have builds for 2.7, 3.5, 3.6.
    * Users of these versions running `pip install grakel` will still recieve a pre-built wheel before this change.
    * As the library is not in active development, this is unlikely to be a problem?
* Moved a lot of the static information from `setup.py` and `setup.cfg` to `pyproject.toml`. 
    *This is a "modern" way to specify meta information about a project, see [PEP 621](https://peps.python.org/pep-0621/) and [setuptools doc page on it](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html).
However, this requires `setuptools>61.0.0` which is only available for `python>=3.6`
  
---
Brief note:

This point above was a bit unnecessary but modern python env/packaging tools such as `poetry` and `flit` use this to quickly generate validate dependancy graphs and find valid ones, using only this meta-data from PyPI. This wasn't possible with `setup.py` specifying requirements as it's a python file and highly un-uniform to parse, while `requirements.txt` does not handle test/dev dependencies. Specifying them in `setup.cfg` would work but it provides little benefit.

In general, the python eco-system is heading towards static build/meta/tooling information in `pyproject.toml`, with python even including a `toml` parser in it's core library soon. It's probably the safest long term bet at the cost of backwards compatibility.

---

### CODE REVIEW

**Comment by ysig** [comment]
Hi @eddiebergman that's great.

Let me address the above points (Todos -> T, Changes -> C):

T1,2: nosetests seems to be the one that changes because of callable see: https://github.com/eddiebergman/GraKeL/actions/runs/3041633008/jobs/4898960711

I have no objection in moving to pytests

T4: Can we replace nosetests with pytest or it is a very big deal?

C2: that's totally fine. As the library is stable for now, we would like to be update with python's changes.

C3: That's great!

**Comment by ysig** [comment]
tldr I think nosetests is the bottleneck right now let's replace it with whatever you think is best (pytest?):
I have no objection to your T1 pipeline testing pipeline.

**Comment by eddiebergman** [comment]
@ysig yup I found them, dont worry :) I ran pytest and nosetest side by side and got the same tests collected and all passed. However pytest additionally reported there's 909 warnings, mostly deprecation warnings from numpy. I can quickly fix those if you like?

**Comment by ysig** [comment]
@eddiebergman

> yup I found them, dont worry :) I ran pytest and nosetest side by side and got the same tests collected and all passed. However pytest additionally reported there's 909 warnings, mostly deprecation warnings from numpy. I can quickly fix those if you like?

Yes that would be great!

**Comment by ysig** [comment]
So what I see as two main errors for now:
1) In  https://github.com/ysig/GraKeL/blob/8500dbf6812116459df308396716b7bcae13ef9c/grakel/tools.py#L175 collections.Hashable should probably be replaced by something that is valid for this version of python.

2) Windows errors seem to be related with cvxopt version - not sure how to approach this for now.

**Comment by eddiebergman** [comment]
Hi @ysig,

Thanks for going in and fixing them :) I'm a bit busy today to do much here but I will fix the wheel builder tests tomorrow or friday!

Best,
Eddi

**Comment by ysig** [comment]
Hi @eddiebergman,

[all are passing right now](https://github.com/eddiebergman/GraKeL/actions/runs/3051878668) except 2 windows processes:
- I installed python 3.10 on my windows 10, but I don't get the same cvxopt related bug so I'm really confused (all tests pass).

Also for codecov reply to my mail so we configure it.
Also can you tell me how I can setup actions to run on my account instead of yours?

For the wheels I see some bugs related with openblas or import k_ij_triangular which is a synonym for the Cython extension not bring installed correctly - not sure why.
It is manually setup using:

```
python setup.py build_ext
python setup.py install
```

Best,
y

**Comment by eddiebergman** [comment]
Just an update, I reran all the tests:
* Seems the wheels are failing due to some import, I will be able to recreate this locally and debug this.
* The two failing windows builds seem to be assertion errors and not cvxopt issues from what I see.
```python
            try: f(x, dy, s)
            except ArithmeticError:
>               raise ValueError("Rank(A) < p or Rank([G; A]) < n")
E               ValueError: Rank(A) < p or Rank([G; A]) < n
```

I imagine this could be due to two cases: **1)** A flaky test where either the code is not fully correct, the test is wrong or numerical instability **2)** Windows 3.9/3.10 ships with a different numerical backend.

I reran the tests, if the same 2 fail then it's very likely **2**, otherwise it is case **1**.

Best,
Eddie

**Comment by eddiebergman** [comment]
I reran the tests and this time it failed for all windows versions, 3.7 to 3.10. My guess is a slight numerical instability or api change with some library on windows which is causing the `ArithmeticError` which is then swallod. I will change it to the second proposed error message to gain some more info.

Failing tests in all windows version seem to do with `lovasz`, whatever that may be.
```
FAILED grakel/tests/test_Kernel.py::test_lovasz_theta - AssertionError: Value...
FAILED grakel/tests/test_Kernel.py::test_lovasz_theta_pd - AssertionError: Va...
FAILED grakel/tests/test_common.py::test_lovasz_theta - ValueError: Rank(A) <...
FAILED grakel/tests/test_kernels.py::test_lovasz_theta - ValueError: Rank(A) ..
```

**Comment by eddiebergman** [comment]
Going to run 100 different seeds on Linux and see if I can reproduce the error, if not, I will submit the same to github actions and see if we get stochastic failures across seeds or stochastic errors across github workers. If it's stochastic across seeds then somehow there is numerical instability with some windows library. If it's stochastic with respect to github action workers then I imagine there is some variance in terms of the packages they download, which might give a further hint.

```python

    # Trying 100 different seeds
    @pytest.mark.parametrize("r", list(range(100)))
    def test_lovasz_theta(r):
        """Random input test for the Lovasz-theta distance kernel."""
        train, test = generate_dataset(n_graphs=50,
                                       r_vertices=(5, 10),
                                       r_connectivity=(0.4, 0.8),
                                       r_weight_edges=(1, 1),
                                       n_graphs_test=20,
                                       random_state=r,
                                       features=None)

        lt_kernel = LovaszTheta(verbose=verbose, normalize=normalize)

        try:
            lt_kernel.fit_transform(train)
            lt_kernel.transform(test)
            assert True
        except Exception as exception:
            assert False, exception
```

**Comment by eddiebergman** [comment]
Seems `100` seeds on Linux still passed everything. From this I'm concluding that it's some numerical instability with Windows that's the prime cause.

Everything stems from `_calculate_lovasz_embeddings` and fails at this [line](https://github.com/eddiebergman/GraKeL/blob/2ffeb1860cd94a9917be606b0d0167e4bc2ac21b/grakel/kernels/lovasz_theta.py#L323) on Windows. I'm going to essentially just print what goes in and see if we can create a minimal working example that at the very least can be posted to `cvxopt` if we can not reproduce it.

```python
def _calculate_lovasz_embeddings_(A):
    """Calculate the lovasz embeddings for the given graph.
    Parameters
    ----------
    A : np.array, ndim=2
        The adjacency matrix.
    Returns
    -------
    lovasz_theta : float
        Returns the lovasz theta number.
    optimization_solution : np.array
        Returns the slack variable solution
        of the primal optimization program.
    """
    if A.shape[0] == 1:
        return 1.0
    else:
        tf_adjm = (np.abs(A) <= min_weight)
        np.fill_diagonal(tf_adjm, False)
        i_list, j_list = np.nonzero(np.triu(tf_adjm.astype(int), k=1))

        nv = A.shape[0]
        ne = len(i_list)

        x_list = list()
        e_list = list()
        for (e, (i, j)) in enumerate(zip(i_list, j_list)):
            e_list.append(int(e)), x_list.append(int(i*nv+j))
            if tf_adjm[i, j]:
                e_list.append(int(e)), x_list.append(int(i*nv+j))

    # Add on the last row, diagonal elements
    e_list = e_list+(nv*[ne])
    x_list = x_list+[int(i*nv + i) for i in range(nv)]

    # initialise g sparse (to values -1, based on two list that
    # define index and one that defines shape
    g_sparse = spmatrix(-1, x_list, e_list, (nv*nv, ne+1))

    # Initialise optimization parameters
    h = matrix(-1.0, (nv, nv))
    c = matrix([0.0]*ne + [1.0])

    # Solve the convex optimization problem
    
    # DEBUGGING
    # =====================
    for label, param in [("c", c), ("g_sparse", g_sparse), ("h", h)]:
        print("== " + label + " ==")
        print(param)
        print("==" * 10)
    # =====================

    sol = sdp(c, Gs=[g_sparse], hs=[h])

    return np.array(sol['ss'][0]), sol['x'][ne]
```

**Comment by eddiebergman** [comment]
Hurray success! The minimal example works and essentially there are failing cases for the `sdp` solver only occuring on windows. https://github.com/eddiebergman/GraKeL/actions/runs/3183531279/jobs/5190864022

**Comment by eddiebergman** [comment]
I can get it down to one single example that fails on Windows but success and mac/linux. Unfortunatly, I don't really know how to proceed from here due to my lack of knowledge of what is being represented and whether this should be solvable. With your go ahead, I would raise an issue with cvxopt github.

```python
# Passed on linux, mac - Fails on Windows (cvxopt 1.3.0)
import pytest
from cvxopt.base import matrix, spmatrix
from cvxopt.solvers import sdp


@pytest.mark.parametrize(
    "nv, ne, e_list, x_list",
    [
        (9,
         15,
         [ 0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6, 7, 7, 8, 8, 9, 9, 10, 10, 11, 11, 12, 12, 13, 13, 14, 14, 15, 15, 15, 15, 15, 15, 15, 15, 15],
         [ 2, 2, 3, 3, 6, 6, 7, 7, 12, 12, 14, 14, 15, 15, 21, 21, 22, 22, 23, 23, 25, 25, 41, 41, 44, 44, 53, 53, 71, 71, 0, 10, 20, 30, 40, 50, 60, 70, 80],
        )
    ],
)
def test_windows_sdp(nv, ne, e_list, x_list) -> None:
    # initialise g sparse (to values -1, based on two list that
    # define index and one that defines shape
    print(nv, ne, e_list, x_list)
    g_sparse = spmatrix(-1, x_list, e_list, (nv * nv, ne + 1))

    # Initialise optimization parameters
    h = matrix(-1.0, (nv, nv))
    c = matrix([0.0] * ne + [1.0])

    # Solve the convex optimization problem
    # Should raise here on windows
    sol = sdp(c, Gs=[g_sparse], hs=[h])
    assert sol is not None

```

**Comment by eddiebergman** [comment]
Okay, at this point I think it's a cvxopt/lapack issue as essentially the failures are stochastic, they seem to fail randomly on windows on different versions, sometimes succeeding for no reason I can discern.

Test 1 (Success on 3.7, 3.8)
* https://github.com/eddiebergman/GraKeL/actions/runs/3183814445/jobs/5191482147

Test 2 (Failed all)
* https://github.com/eddiebergman/GraKeL/actions/runs/3183886579/jobs/5191640524

Assuming the generated `sdp` problem given is correct, my current thoughts on solutions:
* Try be more specific with the `sdp` solver and maybe try out different parameters which lead to something more stable.
* Raise an issue with cvxopt and wait however long that takes.
* Release anyway but not for windows

My preference is definitely for the last one given my reasons for updating GraKeL but I understand this is not ideal. I think all solutions are do-able in parallel, i.e

```python
if windows: 
    solver = partial(sdp, something_windows_specific=...)
    # OR
    raise RuntimeError("LovazsKerenl not available for Windows, see: <issue link>")  
else:
   solver = sdp
   
sol = solver(C, Gs=[g_sparse], hs=[h])
```

I will raise an issue with `cvxopt` anyways with all the details and tag you in there as well in the meantime

Best,
Eddie

**Comment by ysig** [comment]
Hi @eddiebergman 

Sorry for jumping in a bit late. That's amazing work and thank you so much for pinning down this problem.

Lovasz has always been a pain due to numerical instability in windows.

To give you some further context, essentially what happens is the following:
- every kernel is psd. Thus the sdp solver shouldn't fail by the math. It is definetely a numerical error problem and I think it would be great if you raise an issue on cvxopt.
- cvxopt is slow on development so ofc there is no time to waste.
- lovasz is not used that much as a kernel anyway as it's kinda slow so I suggest we comment out the unit tests of windows (for the faulty versions) on our current grakel wheels and through a warning when the kernel is loaded if the distro is windows and python is >= 3.9
- so I propose if possible to help me setup all these other jumbo mumbo (git actions and codecov) so at least all of it is functional and we can push the new version as soon as possible :))

**Comment by ysig** [comment]
Hi @eddiebergman thank you so much!

Two final things I would like you to add before I merge this PR is:

1. Previously we had this script that would upload the build wheels and sdist on pypi https://github.com/ysig/GraKeL/blob/0.1a8/ci_scripts/travis/success.sh can you add this on the wheel builders? I can for sure add the PYPI token 
 
2. it would be great to have a commit keyword to skip uploading 
- equivalently we could have an action to push wheels that can only be triggered by the click of a contributor of the library - me in this case - but I'm not sure whether this is possible.

Lmk if sth doesn't make sense.

**Comment by eddiebergman** [comment]
Hi @ysig,

I [added something](https://github.com/pypa/gh-action-pypi-publish) from Pypa that does it now, I went with the easiest manual solution.
Screenshot_2022-10-06_21-07-35

On the actions tab, you will see the possibility to manually run the `release.yaml` workflow on any branch you have. There will be a checkbox (lower left of image) which will deploy to PyPI if everything runs correctly. The one downside is that this means you will likely run the workflow twice, once on a push to master automatically, and once when you manually select to release. 

If you're not happy with this solution and want it more automated then there are some other possibilities, for example, can trigger only a push to <branch> with a <tag> starting with a particular <substr>. I would advise against trying to automate it since the library is not under active development. I would guess that going to figure out what the automation trigger is not anymore convenient then just triggering it manually from github.

**Comment by eddiebergman** [comment]
Small update:

You can see in this [workflow](https://github.com/eddiebergman/GraKeL/actions/runs/3201746512/jobs/5230481666) in the last step "Upload Release" that it seems to be working as intended. It fails for me as I have no access to PyPI/GraKeL and I also have no access token set. If you look at the log for it, you'll see the wheels and dist that it checks and tries to upload.

I think it's fully mergable then, other than I will resolve the `install.sh` conflict
